[FEATURE][ML] Tools for auto encoding categorical fields

tveasey opened the pull request:
This is helper functionality we'll use for choosing how to encode categorical fields. In particular, it is functionality to 1) compute category frequencies, 2) compute the mean target value for each category, and 3) to compute MICe for hot-one encoded categories and the target variable.

I've split this out to keep the PR smaller.

valeriy42 commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
-                                                    const TSizeVec& columnMask,
-                                                    std::size_t targetColumn) {
+CDataFrameUtils::TSizeDoublePrVecVec
+CDataFrameUtils::categoryMicWithColumnDataFrameInMemory(std::size_t numberThreads,
```
nit: I would arrange the lines in `categoryMicWithColumnDataFrameInMemory` and `categoryMicWithColumnDataFrameOnDisk` in a way that they correspond as much as possible. It may simplify comparison, bug search, or refactoring in the future.

tveasey replied:
I tweaked the function layouts in [b0f154e](https://github.com/elastic/ml-cpp/pull/539/commits/b0f154e35b38c7d66ee559a8a24cc668da8805af). (And also `micWithColumnDataFrameInMemory` and `micWithColumnDataFrameOnDisk`.)

valeriy42 commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
-            samples[slot].second = row[targetColumn];
-        };
-        TRowSampler sampler{numberSamples, onSample};
+        TRowSampler sampler{numberSamples, onSample(i, targetColumn, samples)};
```
nice! :+1:

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -14,20 +14,52 @@
+        if (slot >= samples.size()) {
+            samples.resize(slot + 1, TFloatVec(row.numberColumns()));
+        }
+        row.copyTo(samples[slot].begin());
```
This is sampling the whole row, to avoid having to repeatedly visit the data frame. Here we are completely overwriting the row which was previously in the slot. We use this method to get a very nearly random sample the rows in the data frame with a fixed size in a single pass over them, see `CRandomStreamSampler` for more detail.

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -14,20 +14,52 @@
+using TRowSampler = CSampling::CRandomStreamSampler<TRowRef>;
+
+//! Get a row column sampler.
+auto onSample(std::size_t i, std::size_t targetColumn, TFloatFloatPrVec& samples) {
```
Good suggestion, I've improved the naming in [602d9d6](https://github.com/elastic/ml-cpp/pull/539/commits/602d9d65cbd979bc6ce62f604ea135dec1e0f94d).

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+        return result;
+    }
+
+    auto results = frame.readRows(
```
Done in [25fa629](https://github.com/elastic/ml-cpp/pull/539/commits/25fa6295e9f6dc23993cc290699f6409f5f4e663).

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+    auto results = frame.readRows(
+        numberThreads,
+        core::bindRetrievableState(
+            [&](TDoubleVecVec& counts, TRowItr beginRows, TRowItr endRows) {
```
I think with results renamed this is clearer and so explicitly naming the lambda isn't needed; WDYT?

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+                                    }),
+                     columnMask.end());
+    minimumFrequency = std::max(minimumFrequency,
+                                50.0 / static_cast<double>(frame.numberRows()));
```
I actually decided to remove this constraint when using for category encoding so deleting now instead.

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+        for (std::size_t j = 0; j < means[i].size(); ++j) {
+            result[i][j] = CBasicStatistics::mean(means[i][j]);
+        }
+    }
```
This is a good suggestion.

I think there are enough examples of this now that we can probably have some utilities to reduce the boilerplate code we need to write for these. Let me have a think about this.

My inclination is to make this change in a separate refactor PR.

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+                                static_cast<double>(frame.numberRows());
+            }
+        }
+    }
```
I think [this](https://github.com/elastic/ml-cpp/pull/539#discussion_r303336135) suggestion is the right way to proceed on this. So I'd like to postpone this to a separate PR.

tveasey commented on lib/maths/CDataFrameUtils.cc:
```diff
@@ -122,48 +155,334 @@ bool CDataFrameUtils::columnQuantiles(std::size_t numberThreads,
+
+    if (frame.numberRows() == 0 || columnMask.empty()) {
+        return result;
+    }
```
I think I prefer to have these functions take care of this (rather than every calling site), but definitely agree factoring out is a good idea. This was done in [25fa629](https://github.com/elastic/ml-cpp/pull/539/commits/25fa6295e9f6dc23993cc290699f6409f5f4e663).

tveasey commented:
Thanks for the review @valeriy42, some good suggestions. Regarding almost code duplication I dealt with some low hanging fruit, but my inclination is to see whether we can create some template helpers along the lines of [this](https://github.com/elastic/ml-cpp/pull/539#discussion_r303336135) comment. I'll defer that to a separate PR. Aside from comments related to this point, I think I have now addressed all your suggestions. Can you take another look.
